2024.8 Beta app, unable to log in or sync

### Subject of the issue
Unable to sign into bitwarden app using version 2024.8.0 (beta) on android with 2FA.

Can confirm the webvault works as expected. no issues.

Attempted Yubikey, authenticator app code and email. Each one comes up with an error unable to validate.

Left the beta programme on android and rolled back to 2024.7.1 and all worked as expected.

Logging as a bug in case this is a breaking change on the android app - apologies if should be a discussion

### Deployment environment

* vaultwarden version: 1.32.0

* Install method: docker

* Clients used: android, webvault

* Reverse proxy and version: caddy

* Other relevant details:

### Steps to reproduce
Join android beta, install 2024.8.0 and try to sign in (self hosted)

### Expected behaviour
2FA validates

### Actual behaviour
2FA unable to validate

The only line from the log is: [2024-08-19 10:42:55.196][error][ERROR] 2FA token not provided                                                                                   

Which seems pretty standard, as this appears when the 2FA screen pops up (before submitting)

The exact error on the app is "An error has occurred. Invalid verification code"

Once i attempt to log in, for example using yubi key and i get the above message i immediately get an email to say new device was logged in.

To confirm, this is on a Pixel 8 pro running android 14 (latest security/play system update) and bitwarden app: Version 2024.8.0 com.x8bit.bitwarden.beta

I'm running this side by side now with 2024.7.1 com.xbit.bitwaren which works as expected

> So just to be clear, you do get the window to select a specific 2FA option? And then you are able to provide a code and press the submit button within the client? But then you are denied access??


Thats correct


![1](https://github.com/user-attachments/assets/91abfeb5-56b1-41ee-a8ae-eaa6dfca5b54)
![2](https://github.com/user-attachments/assets/d88333c2-a9e2-4967-a334-cd8284c29463)

```[2024-08-23 11:11:15.130][request][INFO] GET /api/devices/knowndevice
[2024-08-23 11:11:15.133][response][INFO] (get_known_device) GET /api/devices/knowndevice => 200 OK
[2024-08-23 11:11:26.482][request][INFO] POST /identity/accounts/prelogin
[2024-08-23 11:11:26.483][response][INFO] (prelogin) POST /identity/accounts/prelogin => 200 OK
[2024-08-23 11:11:27.366][request][INFO] POST /identity/connect/token
[2024-08-23 11:11:27.499][error][ERROR] 2FA token not provided
[2024-08-23 11:11:27.499][response][INFO] (login) POST /identity/connect/token => 400 Bad Request
[2024-08-23 11:11:44.628][request][INFO] POST /identity/connect/token
[2024-08-23 11:11:44.783][vaultwarden::api::identity][INFO] User ABC@DEF.XYZ logged in successfully. IP: XXX.XXX.XXX.XXX
[2024-08-23 11:11:44.784][response][INFO] (login) POST /identity/connect/token => 200 OK

> > So just to be clear, you do get the window to select a specific 2FA option? And then you are able to provide a code and press the submit button within the client? But then you are denied access??
> 
> Thats correct
> 
> ![1](https://private-user-images.githubusercontent.com/555893/360864599-91abfeb5-56b1-41ee-a8ae-eaa6dfca5b54.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU1NDg4ODQsIm5iZiI6MTcyNTU0ODU4NCwicGF0aCI6Ii81NTU4OTMvMzYwODY0NTk5LTkxYWJmZWI1LTU2YjEtNDFlZS1hOGFlLWVhYTZkZmNhNWI1NC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTA1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkwNVQxNTAzMDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03MWI5YmNlYTIwMTJhNzEyYzEzZDNjMzJlODIzMjg2ZGRjMzM2YmNiODIzYzg4YTVhMzJmYTIxYjU1NDk1Y2RkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.ZrDhvqn-ukvqfs4afxPFt8yfrLO2tyKh-d8bNn26nwY) ![2](https://private-user-images.githubusercontent.com/555893/360864644-d88333c2-a9e2-4967-a334-cd8284c29463.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU1NDg4ODQsIm5iZiI6MTcyNTU0ODU4NCwicGF0aCI6Ii81NTU4OTMvMzYwODY0NjQ0LWQ4ODMzM2MyLWE5ZTItNDk2Ny1hMzM0LWNkODI4NGMyOTQ2My5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTA1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkwNVQxNTAzMDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mMzY0YzVhMjI5ZjZlYmM2MDAyMDNhZWM1NjJkNWUzYjc0NzlhYjk2MWU3MTY0YzIxYzhiZDhlYzcyOWEyMDg5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.2pNE9a_uq7WeKqHJLb4ToxFcJA9NU-jBx1irbrBO2jE)
> 
> **Newest Android App** 2FA failed
> 
> ```
> [2024-08-23 11:11:15.130][request][INFO] GET /api/devices/knowndevice
> [2024-08-23 11:11:15.133][response][INFO] (get_known_device) GET /api/devices/knowndevice => 200 OK
> [2024-08-23 11:11:26.482][request][INFO] POST /identity/accounts/prelogin
> [2024-08-23 11:11:26.483][response][INFO] (prelogin) POST /identity/accounts/prelogin => 200 OK
> [2024-08-23 11:11:27.366][request][INFO] POST /identity/connect/token
> [2024-08-23 11:11:27.499][error][ERROR] 2FA token not provided
> [2024-08-23 11:11:27.499][response][INFO] (login) POST /identity/connect/token => 400 Bad Request
> [2024-08-23 11:11:44.628][request][INFO] POST /identity/connect/token
> [2024-08-23 11:11:44.783][vaultwarden::api::identity][INFO] User ABC@DEF.XYZ logged in successfully. IP: XXX.XXX.XXX.XXX
> [2024-08-23 11:11:44.784][response][INFO] (login) POST /identity/connect/token => 200 OK
> ```
> 
> **2024.7.2 (18910) beta** 2FA works
> 
> ```
> [2024-08-23 11:20:26.165][request][INFO] GET /api/devices/knowndevice
> [2024-08-23 11:20:26.167][response][INFO] (get_known_device) GET /api/devices/knowndevice => 200 OK
> [2024-08-23 11:20:36.564][request][INFO] POST /identity/accounts/prelogin
> [2024-08-23 11:20:36.566][response][INFO] (prelogin) POST /identity/accounts/prelogin => 200 OK
> [2024-08-23 11:20:37.660][request][INFO] POST /identity/connect/token
> [2024-08-23 11:20:37.793][error][ERROR] 2FA token not provided
> [2024-08-23 11:20:37.794][response][INFO] (login) POST /identity/connect/token => 400 Bad Request
> [2024-08-23 11:20:45.379][request][INFO] POST /identity/connect/token
> [2024-08-23 11:20:48.278][vaultwarden::api::identity][INFO] User ABC@DEF.XYZ logged in successfully. IP: XXX.XXX.XXX.XXX
> [2024-08-23 11:20:48.279][response][INFO] (login) POST /identity/connect/token => 200 OK
> [2024-08-23 11:20:50.632][request][INFO] GET /api/accounts/revision-date
> [2024-08-23 11:20:50.634][response][INFO] (revision_date) GET /api/accounts/revision-date => 200 OK
> [2024-08-23 11:20:50.860][request][INFO] GET /api/sync
> [2024-08-23 11:20:50.908][response][INFO] (sync) GET /api/sync?<data..> => 200 OK
> ```

I have exactly the same problem.
Except that both the server and the app are freshly installed.

--- a/src/api/core/accounts.rs
+++ b/src/api/core/accounts.rs
@@ -490,7 +490,7 @@ async fn post_rotatekey(data: Json<KeyData>, headers: Headers, mut conn: DbConn,
     // Bitwarden does not process the import if there is one item invalid.
     // Since we check for the size of the encrypted note length, we need to do that here to pre-validate it.
     // TODO: See if we can optimize the whole cipher adding/importing and prevent duplicate code and checks.
-    Cipher::validate_notes(&data.ciphers)?;
+    Cipher::validate_cipher_data(&data.ciphers)?;
 
     let user_uuid = &headers.user.uuid;
 

--- a/src/api/core/ciphers.rs
+++ b/src/api/core/ciphers.rs
@@ -233,7 +233,7 @@ pub struct CipherData {
     favorite: Option<bool>,
     reprompt: Option<i32>,
 
-    password_history: Option<Value>,
+    pub password_history: Option<Value>,
 
     // These are used during key rotation
     // 'Attachments' is unused, contains map of {id: filename}
@@ -563,7 +563,7 @@ async fn post_ciphers_import(
     // Bitwarden does not process the import if there is one item invalid.
     // Since we check for the size of the encrypted note length, we need to do that here to pre-validate it.
     // TODO: See if we can optimize the whole cipher adding/importing and prevent duplicate code and checks.
-    Cipher::validate_notes(&data.ciphers)?;
+    Cipher::validate_cipher_data(&data.ciphers)?;
 
     // Read and create the folders
     let existing_folders: Vec<String> =

--- a/src/api/core/organizations.rs
+++ b/src/api/core/organizations.rs
@@ -1596,7 +1596,7 @@ async fn post_org_import(
     // Bitwarden does not process the import if there is one item invalid.
     // Since we check for the size of the encrypted note length, we need to do that here to pre-validate it.
     // TODO: See if we can optimize the whole cipher adding/importing and prevent duplicate code and checks.
-    Cipher::validate_notes(&data.ciphers)?;
+    Cipher::validate_cipher_data(&data.ciphers)?;
 
     let mut collections = Vec::new();
     for coll in data.collections {

--- a/src/db/models/cipher.rs
+++ b/src/db/models/cipher.rs
@@ -79,19 +79,39 @@ impl Cipher {
         }
     }
 
-    pub fn validate_notes(cipher_data: &[CipherData]) -> EmptyResult {
+    pub fn validate_cipher_data(cipher_data: &[CipherData]) -> EmptyResult {
         let mut validation_errors = serde_json::Map::new();
         let max_note_size = CONFIG._max_note_size();
         let max_note_size_msg =
             format!("The field Notes exceeds the maximum encrypted value length of {} characters.", &max_note_size);
         for (index, cipher) in cipher_data.iter().enumerate() {
+            // Validate the note size and if it is exceeded return a warning
             if let Some(note) = &cipher.notes {
                 if note.len() > max_note_size {
                     validation_errors
                         .insert(format!("Ciphers[{index}].Notes"), serde_json::to_value([&max_note_size_msg]).unwrap());
                 }
             }
+
+            // Validate the password history if it contains `null` values and if so, return a warning
+            if let Some(Value::Array(password_history)) = &cipher.password_history {
+                for pwh in password_history {
+                    if let Value::Object(pwo) = pwh {
+                        if pwo.get("password").is_some_and(|p| !p.is_string()) {
+                            validation_errors.insert(
+                                format!("Ciphers[{index}].Notes"),
+                                serde_json::to_value([
+                                    "The password history contains a `null` value. Only strings are allowed.",
+                                ])
+                                .unwrap(),
+                            );
+                            break;
+                        }
+                    }
+                }
+            }
         }
+
         if !validation_errors.is_empty() {
             let err_json = json!({
                 "message": "The model state is invalid.",
@@ -153,27 +173,39 @@ impl Cipher {
             .as_ref()
             .and_then(|s| {
                 serde_json::from_str::<Vec<LowerCase<Value>>>(s)
-                    .inspect_err(|e| warn!("Error parsing fields {:?}", e))
-                    .ok()
-            })
-            .map(|d| d.into_iter().map(|d| d.data).collect())
-            .unwrap_or_default();
-        let password_history_json: Vec<_> = self
-            .password_history
-            .as_ref()
-            .and_then(|s| {
-                serde_json::from_str::<Vec<LowerCase<Value>>>(s)
-                    .inspect_err(|e| warn!("Error parsing password history {:?}", e))
+                    .inspect_err(|e| warn!("Error parsing fields {e:?} for {}", self.uuid))
                     .ok()
             })
             .map(|d| d.into_iter().map(|d| d.data).collect())
             .unwrap_or_default();
 
+        let password_history_json: Vec<_> = self
+            .password_history
+            .as_ref()
+            .and_then(|s| {
+                serde_json::from_str::<Vec<LowerCase<Value>>>(s)
+                    .inspect_err(|e| warn!("Error parsing password history {e:?} for {}", self.uuid))
+                    .ok()
+            })
+            .map(|d| {
+                // Check every password history item if they are valid and return it.
+                // If a password field has the type `null` skip it, it breaks newer Bitwarden clients
+                d.into_iter()
+                    .filter_map(|d| match d.data.get("password") {
+                        Some(p) if p.is_string() => Some(d.data),
+                        _ => None,
+                    })
+                    .collect()
+            })
+            .unwrap_or_default();
+
         // Get the type_data or a default to an empty json object '{}'.
         // If not passing an empty object, mobile clients will crash.
-        let mut type_data_json = serde_json::from_str::<LowerCase<Value>>(&self.data)
-            .map(|d| d.data)
-            .unwrap_or_else(|_| Value::Object(serde_json::Map::new()));
+        let mut type_data_json =
+            serde_json::from_str::<LowerCase<Value>>(&self.data).map(|d| d.data).unwrap_or_else(|_| {
+                warn!("Error parsing data field for {}", self.uuid);
+                Value::Object(serde_json::Map::new())
+            });
 
         // NOTE: This was marked as *Backwards Compatibility Code*, but as of January 2021 this is still being used by upstream
         // Set the first element of the Uris array as Uri, this is needed several (mobile) clients.
@@ -189,10 +221,13 @@ impl Cipher {
 
         // Fix secure note issues when data is invalid
         // This breaks at least the native mobile clients
-        if self.atype == 2
-            && (self.data.is_empty() || self.data.eq("{}") || self.data.to_ascii_lowercase().eq("{\"type\":null}"))
-        {
-            type_data_json = json!({"type": 0});
+        if self.atype == 2 {
+            match type_data_json {
+                Value::Object(ref t) if t.get("type").is_some_and(|t| t.is_number()) => {}
+                _ => {
+                    type_data_json = json!({"type": 0});
+                }
+            }
         }
 
         // Clone the type_data and add some default value.

